### PR TITLE
Implemented success response with an empty array

### PR DESCRIPTION
**Pull request by JacekAndrzejewski**
1. Pulled out default success response data to a field
2. Added fluent method for setting default success response data
3. Added tests for usage of new method
4. Added missing tests for null values in respondWithSuccess, respondCreated, respondNoContent and fixed problems
    - null values were allowed both in types and phpdocs, yet exceptions were thrown when passed, this is fixed and null values are returned as empty arrays
5. Added missing tests for Arrayable, JsonSerializable and Collection and their empty variants
6. Strengthened tests to use strict assertions

### CODE REVIEW

File: src/ApiResponseHelpers.php [hunk 1]
```diff
@@ -14,6 +14,9 @@
 
 trait ApiResponseHelpers
 {
+
+    private ?array $_api_helpers_defaultSuccessData = ['success' => true];
```
Comment (JacekAndrzejewski) [inline src/ApiResponseHelpers.php:17]: That name probably won't be used by any project anywhere, but I'm glad to hear suggestions

File: src/ApiResponseHelpers.php [hunk 2]
```diff
@@ -33,17 +36,23 @@ public function respondNotFound(
     /**
      * @param array|Arrayable|JsonSerializable|null $contents
      */
-    public function respondWithSuccess($contents = []): JsonResponse
+    public function respondWithSuccess($contents = null): JsonResponse
     {
-        $contents = $this->morphToArray($contents);
+        $contents = $this->morphToArray($contents) ?? [];
```
Comment (JacekAndrzejewski) [inline src/ApiResponseHelpers.php:40]: Change of default value won't impact calls without arguments, but will allow to cleanly change passed nulls to empty arrays when someone does pass null. Empty arrays will stay the same as before.
This change is in respondNoContent and respondCreated too.

File: tests/ResponseTest.php [hunk 3]
```diff
@@ -35,8 +36,21 @@ public function testResponses(string $method, array $args, int $code, array $dat
         /** @var \Illuminate\Http\JsonResponse $response */
         $response = call_user_func_array([$this->service, $method], $args);
         self::assertInstanceOf(JsonResponse::class, $response);
-        self::assertEquals($code, $response->getStatusCode());
-        self::assertEquals($data, $response->getData(true));
+        self::assertSame($code, $response->getStatusCode());
```
Comment (JacekAndrzejewski) [inline tests/ResponseTest.php:39]: Ummm yeah, that's not what happened sir Git

File: tests/ResponseTest.php [hunk 4]
```diff
@@ -74,6 +88,13 @@ public function basicResponsesDataProvider(): array
             ['success' => true],
           ],
 
+          'respondWithSuccess(), null response data' => [
+            'respondWithSuccess',
+            [null],
+            Response::HTTP_OK,
+            ['success' => true],
+          ],
```
Comment (JacekAndrzejewski) [inline tests/ResponseTest.php:96]: This case was missing and failed before my changes. Null is allowed based on phpdocs and type hints. There are few similar cases down below.

File: tests/ResponseTest.php [hunk 5]
```diff
@@ -95,6 +116,69 @@ public function basicResponsesDataProvider(): array
...
+              new class implements Arrayable {
+                public function toArray()
+                {
+                  return [];
+                }
+              },
+            ],
+            Response::HTTP_OK,
+            ['success' => true]
+          ],
```
Comment (JacekAndrzejewski) [inline tests/ResponseTest.php:152]: Those didn't fail before my changes, however if you would break morphToArray on purpose for Arrayable only then that would not be caught by tests. That's why I added those.

File: tests/ResponseTest.php [hunk 6]
```diff
@@ -35,8 +36,21 @@ public function testResponses(string $method, array $args, int $code, array $dat
...
+    /**
+     * @dataProvider successDefaultsDataProvider
+     * @throws JsonException
+     */
+    public function testSuccessResponseDefaults(?array $default, $args, int $code, array $data): void
+    {
+        $response = $this->service->setDefaultSuccessResponse($default)->respondWithSuccess($args);
+        self::assertInstanceOf(JsonResponse::class, $response);
+        self::assertSame($code, $response->getStatusCode());
+        self::assertSame($data, $response->getData(true));
```
Comment (JacekAndrzejewski) [inline tests/ResponseTest.php:53]: We could technically merge this and other test method, but it would increase complexity too much IMO. It's already complex enough.

File: src/ApiResponseHelpers.php [hunk 7]
```diff
@@ -33,17 +36,23 @@ public function respondNotFound(
...
 
         $data = [] === $contents
-            ? ['success' => true]
+            ? $this->_api_helpers_defaultSuccessData
             : $contents;
 
         return $this->apiResponse($data);
     }
 
+    public function setDefaultSuccessResponse(?array $content = null)
```
Comment (JacekAndrzejewski) [inline src/ApiResponseHelpers.php]: We could add a `self` or `static` phpdoc, however from my tests this doesn't help PHPStorm realize that this is fluent method returning class using this trait. I were unable to make it understand that.
But it realizes it returns object with trait methods so it's good enough. If anyone has idea how to make this better then give this a shot.

**Comment by JacekAndrzejewski** [comment]
This should solve everything discussed in #13.
It should also be compatible with #15, with possibly minor changes.
It should be backwards compatible, one place where it's not is that passing null no longer throws an exception. If someone called methods with null and used try-catch and ignored exceptions it WOULD be a BC. However I think we can allow that case, because if someone really does that with methods returning responses then they have worse problems with their code.

I tried explaining anything, if something is hard to understand or needs fixing please ask.
@ultrono Please close conversations where you have no questions and run workflows if all is ok.

**Comment by ultrono** [comment]
Great PR and explanation @JacekAndrzejewski, much appreciated.  Have finally got my schedule under control and will look at getting this merged asap.

File: src/ApiResponseHelpers.php (hunk 2 above)
Comment (ultrono) [reply src/ApiResponseHelpers.php:40]: Yep kudos there, this was exactly my plan as to avoid a full breaking change 👍

File: src/ApiResponseHelpers.php (hunk 7 above)
Comment (ultrono) [reply src/ApiResponseHelpers.php]: I'll take a look at that, usually `self` has always been enough. However, as this is via a trait, as you've discovered, that may change.

**Comment by ultrono** [comment]
After merging this in (everything looks great from an initial look) I think a small update the readme would be beneficial too.

File: src/ApiResponseHelpers.php (hunk 7 above)
Comment (ultrono) [reply src/ApiResponseHelpers.php]: PHPStorm was nagging about no return type and suggested `self`. I've added that for good measure.

**Comment by ultrono** [comment]
@JacekAndrzejewski Thanks for this, great PR 👍

**Comment by JacekAndrzejewski** [comment]
@ultrono By the way, could you label this with `hacktoberfest-accepted`? I want to plant a tree :D

**Comment by ultrono** [comment]
> 
> 
> @ultrono By the way, could you label this with `hacktoberfest-accepted`? I want to plant a tree :D

No probs, all added.
